### ISSUE TITLE
Implement Hamming Loss for classifier performance

### ISSUE THREAD

**Issue by peterjc** [peterjc/thapbi-pict#71]
This is a popular metric in multi-class classifier literature, see e.g.

Sokolova and Lapalme (2009) A systematic analysis of performance measures for classification tasks
https://doi.org/10.1016/j.ipm.2009.03.002

Pereira *et al.* (2018) Correlation analysis of performance measures for multi-label classification
https://doi.org/10.1016/j.ipm.2018.01.002

**Comment by peterjc** [comment]
Results from branch https://github.com/peterjc/thapbi-pict/tree/hamming_loss using 70 single isolates samples (covering 30 species) on a control plate and various test predictions:

```
$ thapbi_pict assess . -m known | head -n 2
Assessed known vs known in 70 files (30 species; 174 sequence entries)
#Species	TP	FP	FN	TN	sensitivity	specificity	precision	F1	Hamming-loss
OVERALL	174	0	0	0	1.00	0.00	1.00	1.00	0.0000

$ thapbi_pict assess . -m nope | head -n 2
Assessed nope vs known in 70 files (30 species; 174 sequence entries)
#Species	TP	FP	FN	TN	sensitivity	specificity	precision	F1	Hamming-loss
OVERALL	0	0	174	0	0.00	0.00	0.00	0.00	0.0333

$ thapbi_pict assess . -m infestans | head -n 2
Assessed infestans vs known in 70 files (30 species; 174 sequence entries)
#Species	TP	FP	FN	TN	sensitivity	specificity	precision	F1	Hamming-loss
OVERALL	30	144	144	0	0.17	0.00	0.17	0.17	0.0552

$ thapbi_pict assess . -m big3 | head -n 2
Assessed big3 vs known in 70 files (30 species; 174 sequence entries)
#Species	TP	FP	FN	TN	sensitivity	specificity	precision	F1	Hamming-loss
OVERALL	76	446	98	0	0.44	0.00	0.15	0.22	0.1042

$ thapbi_pict assess . -m everything | head -n 2
Assessed everything vs known in 70 files (30 species; 174 sequence entries)
#Species	TP	FP	FN	TN	sensitivity	specificity	precision	F1	Hamming-loss
OVERALL	174	5046	0	0	1.00	0.00	0.03	0.06	0.9667
```

* ``known`` = prediction is the expected answer, perfect scores, Hamming Loss 0.0000
* ``nope`` = prediction is always no *Phytophthora* at all, Hamming Loss 0.0333
* ``infestans`` = prediction is always *P. infestans*, Hamming Loss 0.0552
* ``big3`` = prediction is always three most common *Phytophthora* in this set, Hamming Loss 0.1042
* ``everything`` = prediction is always all 30  *Phytophthora* species, Hamming Loss 0.9667

i.e. As expected get a range of Hamming Loss from zero to (almost) one on this test set.

We get lower Hamming Loss for the ``identity``, ``swarm`` and ``blast`` classifiers (i.e. they do better than these artificial test cases).

Currently this implementation is using the union of expected and predicted species as the number of classes - arguably should use the maximum of the expected species and the total number of species in the classifier's database, which would make these actual predictor score scores lower overall.

**Comment by peterjc** [comment]
Fixed in v0.0.7

**Comment by peterjc** [comment]
See work in v0.0.8 and #75  to include all the possible species. Because I now track how many species there could have been predictions made for (in order to report full FN count), I now have a larger divisor in the Hamming Loss calculation, making the reported losses smaller.

**Comment by peterjc** [comment]
Example with 5 species (A, B, C, D and E) and 6 samples. Here I am using my large TN counts approach as in v0.0.8:

- Sample 1: expected A, predicted A, gives 1 TP, 4 TN.
- Sample 2: expected B, predicted B, gives 1 TP, 4TN.
- Sample 3: expected C, predicted B and C, gives 1 TP (the C), 1 FP (the B), 3 TN (A, D and E).
- Sample 4: expected D, predicted C, gives 1 FP (the C), 1 FN (missing D), 3 TN (A, B and E).
- Sample 5: expected E, predicted E, gives 1 TP, 4 TN.
- Sample 6: expected A and E, predicted A only, gives 1 TP (the A), 1 FN (missing E), 3 TN (B, C and D.
- Totals: 5 TP, 2 FP, 2 FN, 21 TN

With this interpretation, for each sample TP+FP+FN+TN = number of species = 5. Therefore the grand total TP+FP+FN+TN = number of species * number of samples = 5 * 6 = 30.

My Hamming Loss calculation is total FP + total FN, divided by number of samples times number of species:

(total FP + FN) / (#species * #samples) = (2 + 2) / (6 * 5) = 4/30 = 0.13333333333333333

i.e. Number of mistakes (FP + FN) per prediction for each sample-species combination.

It seems earlier versions of Scikit-Learn’s Hamming Loss function accepted lists of lists as a way to express multilabel classification results (e.g. this sample is species A only, this sample is species A and E, etc), which is close to how I am storing classifier output internally. They instead now have a helper function to turn class lists into their multilabel format, a (samples x classes) binary matrix indicating the presence of a class label.

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.html

```python
>>> from sklearn.metrics import hamming_loss
>>> from sklearn.preprocessing import MultiLabelBinarizer
>>> relabel = MultiLabelBinarizer(["A", "B", "C", "D", "E"])
>>> expt = relabel.fit_transform([["A"], ["B"], ["C"], ["D"], ["E"], ["A", "E"]])
>>> pred = relabel.fit_transform([["A"], ["B"], ["B", "C"], ["C"], ["E"], ["A"]])
>>> expt
array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 1]])
>>> pred
array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 1, 1, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]])
>>> hamming_loss(expt, pred)
0.13333333333333333
>>> 4/(5*6)
0.13333333333333333
```

If we modify this to say actually the classifier knows about 8 species, "A" to "E" (the 5 present in the expected values) plus "F", "G" and "H", the numbers change as follows:

```python
>>> from sklearn.metrics import hamming_loss
>>> from sklearn.preprocessing import MultiLabelBinarizer
>>> relabel = MultiLabelBinarizer(["A", "B", "C", "D", "E", "F", "G", "H"])
>>> expt = relabel.fit_transform([["A"], ["B"], ["C"], ["D"], ["E"], ["A", "E"]])
>>> pred = relabel.fit_transform([["A"], ["B"], ["B", "C"], ["C"], ["E"], ["A"]])
>>> expt
array([[1, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 0],
       [0, 0, 1, 0, 0, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 0, 0],
       [0, 0, 0, 0, 1, 0, 0, 0],
       [1, 0, 0, 0, 1, 0, 0, 0]])
>>> pred
array([[1, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 0, 0, 0, 0],
       [0, 0, 1, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 1, 0, 0, 0],
       [1, 0, 0, 0, 0, 0, 0, 0]])
>>> hamming_loss(expt, pred)
0.08333333333333333
>>> 4/(8*6)
0.08333333333333333
```

Again, this matches my Hamming Loss calculation = (total FP + FN) / (#species * #samples) = (2 + 2) / (8 * 5) = 4/48 = 0.08333333333333333

So, while this all seems to be consistent, it means Hamming Loss as implemented may not be as useful a metric as we hoped for.

**Comment by peterjc** [comment]
Now for an example were each sample expectation and prediction is single valued, again from a pool of 5 species:

```python
>>> from sklearn.metrics import hamming_loss
>>> from sklearn.preprocessing import MultiLabelBinarizer
>>> relabel = MultiLabelBinarizer(["A", "B", "C", "D", "E"])  # 5 speces
>>> expt = relabel.fit_transform([["A"], ["B"], ["C"], ["D"], ["E"], ["A"]])
>>> pred = relabel.fit_transform([["A"], ["B"], ["B"], ["C"], ["E"], ["A"]])
>>> hamming_loss(expt, pred)
0.13333333333333333
```

My calculation is (2 FP + 2 FN) / (5 species * 6 samples) = 4 / 30 = 0.13333333333333333

```python
>>> expt
array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]])
>>> pred
array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]])
>>> np.bitwise_xor(expt, pred)
array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 1, 1, 0, 0],
       [0, 0, 1, 1, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])
>>> np.bitwise_xor(expt, pred).sum()
4
```

Note that SciKit will accept this non-nested input:


```
>>> relabel.fit_transform([["A"], ["B"], ["C"], ["D"], ["E"], ["A"]])
array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]])
>>> relabel.fit_transform(["A", "B", "C", "D", "E", "A"])
array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]])
```

Now, if we give these single-values classes directly to the Hamming Loss function...

```
>>> hamming_loss([["A"], ["B"], ["C"], ["D"], ["E"], ["A"]], [["A"], ["B"], ["B"], ["C"], ["E"], ["A"]])
0.3333333333333333
>>> hamming_loss(["A", "B", "C", "D", "E", "A"], ["A", "B", "B", "C", "E", "A"])
0.3333333333333333
```

This fits with a binary interpretation, 2 wrong out of 6 predictions, 2/6 = 0.3333333333333333
